pkg/discovery/processor/cluster_processor.go

```diff
@@ -8,111 +8,134 @@ import (
-	ClusterInfoScraper *cluster.ClusterScraper
-	NodeScrapper       *kubeclient.KubeletClient
+	clusterInfoScraper cluster.ClusterScraperInterface
+	nodeScrapper       kubeclient.KubeHttpClientInterface
 }
 
-// Query the Kubernetes API Server to get the cluster nodes and namespaces.
-func (processor *ClusterProcessor) ConnectCluster(checkAllNodes bool) (*repository.KubeCluster, error) {
-	svcID, err := processor.ClusterInfoScraper.GetKubernetesServiceID()
+func NewClusterProcessor(kubeClient *cluster.ClusterScraper, kubeletClient *kubeclient.KubeletClient) *ClusterProcessor {
+	if kubeClient == nil {
+		glog.Errorf("Null kubeclient while creating cluster processor")
+		return nil
+	}
+	if kubeletClient == nil {
+		glog.Errorf("Null kubeletclient while creating cluster processor")
+		return nil
+	}
+	clusterProcessor := &ClusterProcessor{
+		clusterInfoScraper: kubeClient,
+		nodeScrapper:       kubeletClient,
+	}
+	return clusterProcessor
+}
+
+// Connects to the Kubernetes API Server and the nodes in the cluster.
+// If connection is successful a KubeCluster entity to represent the cluster is created
+func (processor *ClusterProcessor) ConnectCluster() (*repository.KubeCluster, error) {
+	if processor.clusterInfoScraper == nil || processor.nodeScrapper == nil {
+		return nil, fmt.Errorf("Null kubernetes cluster or node client")
+	}
+
+	svcID, err := processor.clusterInfoScraper.GetKubernetesServiceID()
 	if err != nil {
 		return nil, fmt.Errorf("Cannot obtain service ID for cluster %s\n", err)
 	}
 	kubeCluster := repository.NewKubeCluster(svcID)
+	glog.V(4).Infof("Created cluster %s\n", svcID)
 
 	// Nodes
-	err = processor.ConnectToNodes(checkAllNodes)
+	err = processor.connectToNodes()
+	if err != nil {
+		return nil, err
+	}
 
 	return kubeCluster, err
 }
 
-func (processor *ClusterProcessor) ConnectToNodes(checkAllNodes bool) error {
-	nodeList, err := processor.ClusterInfoScraper.GetAllNodes()
+// Connects to all nodes in the cluster.
+// Obtains the CPU Frequency of the node to determine the node availability and accessibility.
+func (processor *ClusterProcessor) connectToNodes() error {
+	nodeList, err := processor.clusterInfoScraper.GetAllNodes()
 	if err != nil {
 		return fmt.Errorf("Error getting nodes for cluster : %s\n", err)
 	}
 	glog.V(2).Infof("There are %d nodes\n", len(nodeList))
 
-	var connectionErrors []error
+	var connectionErrors []string
 	var connectionMsgs []string
 
-	n := rand.Int() % len(nodeList)
-	for idx, node := range nodeList {
-		if !checkAllNodes && idx != n {
-			continue
-		}
-		ip := repository.ParseNodeIP(node)
-		kc := processor.NodeScrapper
-		cpuFreq, err := kc.GetMachineCpuFrequency(ip)
+	for _, node := range nodeList {
+		nodeCpuFrequency, err := checkNode(node, processor.nodeScrapper)
 
 		if err != nil {
-			connectionErrors = append(connectionErrors, fmt.Errorf("%s:%s", node.Name, err))
+			connectionErrors = append(connectionErrors, fmt.Sprintf("%s:%s", node.Name, err))
```

Comment:
If we fail connecting to the node, it will be helpful to know the node status. So, we can also add node.Status in the message.